apps/opencs/view/render/instancemode.cpp

```diff
@@ -681,3 +698,180 @@ void CSVRender::InstanceMode::deleteSelectedInstances(bool active)
+    for(osg::ref_ptr<TagBase> tag: selection)
+    {
+        if (CSVRender::ObjectTag *objectTag = dynamic_cast<CSVRender::ObjectTag *> (tag.get()))
+        {
+            osg::ref_ptr<osg::Group> objectNodeWithGUI = objectTag->mObject->getRootNode();
+            osg::ref_ptr<osg::Group> objectNodeWithoutGUI = objectTag->mObject->getBaseNode();
+
+            osg::ComputeBoundsVisitor computeBounds;
+            computeBounds.setTraversalMask(SceneUtil::Mask_EditorReference);
+            objectNodeWithoutGUI->accept(computeBounds);
+            osg::BoundingBox bounds = computeBounds.getBoundingBox();
+            float boundingBoxOffset = 0.0f;
+            if (bounds.valid()) boundingBoxOffset = bounds.zMin();
+
+            objectHeights.emplace_back(boundingBoxOffset);
+            oldMasks.emplace_back(objectNodeWithGUI->getNodeMask());
+
+            objectNodeWithGUI->setNodeMask(SceneUtil::Mask_Disabled);
+        }
+    }
+
+    try
+    {
+        switch (dropMode)
+        {
+            case Terrain:
+            case Collision:
+            {
+                float smallestDropHeight = std::numeric_limits<float>::max();
+                int counter = 0;
+                    for(osg::ref_ptr<TagBase> tag: selection)
+                        if (CSVRender::ObjectTag *objectTag = dynamic_cast<CSVRender::ObjectTag *> (tag.get()))
+                        {
+                            float thisDrop = getDropHeight(dropMode, objectTag->mObject, objectHeights[counter]);
+                            if (thisDrop < smallestDropHeight) smallestDropHeight = thisDrop;
+                            counter++;
+                        }
+                    for(osg::ref_ptr<TagBase> tag: selection)
+                        if (CSVRender::ObjectTag *objectTag = dynamic_cast<CSVRender::ObjectTag *> (tag.get()))
+                        {
+                            objectTag->mObject->setEdited (Object::Override_Position);
+                            ESM::Position position = objectTag->mObject->getPosition();
+                            position.pos[2] -= smallestDropHeight;
+                            objectTag->mObject->setPosition(position.pos);
+                            objectTag->mObject->apply (macro);
+                        }
+            }
+                break;
+
+            case Terrain_sep:
+            case Collision_sep:
+            {
+                int counter = 0;
+                for(osg::ref_ptr<TagBase> tag: selection)
+                    if (CSVRender::ObjectTag *objectTag = dynamic_cast<CSVRender::ObjectTag *> (tag.get()))
+                    {
+                        dropInstance(dropMode, objectTag->mObject, objectHeights[counter]);
+                        objectTag->mObject->apply (macro);
+                        counter++;
+                    }
+            }
+                break;
+            default:
+                break;
+        }
+    }
+    catch (const std::exception& e)
```

Comment:
This is inconsistent with how we handle exceptions. Typically we let them propagate to the event loop, where they are handled and logged (not 100% sure of the logged part). We should keep this consistent (in case we want to change the error handling/logging behaviour at some point).

Reply:
I'm not quite sure what you mean. This is almost the same as https://github.com/OpenMW/openmw/blob/master/apps/opencs/view/render/scenewidget.cpp#L112 . There are other cases where a Qt signal is emitted at exception, is this what you meant?

Reply:
The code you pointed out is a special case, because it sits in a destructor. Destructors should not throw exceptions.

Our OpenMW-CS exception handling code is located in Application::notify (main.cpp). This is the central place where all exception handling happens. Code in event or signal handlers should not do their own exception handling other than local cleanup (if required). In this case the exception must be rethrown (via a simple throw; statement) so it can reach the main exception handling code.
Of course ideally code would just utilise RAII, in which case the cleanup would happen automatically without a try catch clause.

Reply:
I see, but.. I'm not really sure where the exception could occur anyway, and I've tried to use automatic clean-up whenever possible. Sure there are various OpenMW classes and osg smart pointers whose behavior might be hard to predict, but then, I don't see why exactly this code would require try-catch, and why other osg-using code wouldn't need that.

But I assume you have a reason to suggest this, so, instead of:

```
    catch (const std::exception& e)
    {
        Log(Debug::Error) << "Error in dropping instance: " << e.what();
    }
```

should I do this?
```
    catch (const std::exception& e)
    {
        throw std::runtime_error ("Error in dropping instance.");
    }
```

Reply:
It is reasonable to assume that in C++ (almost) anything can throw (almost) anything, unless you have a nothrow specifier, of course. Just because the code does not throw now doesn't mean it won't after a future change. That is a question of having maintainable and robust code.

Unfortunately the solution you posted won't work. First, you do not reach the cleanup code in case of an exception. And second, you throw the wrong exception. You should re-throw by using a

`
throw;
`

statement; that is, if you do not utilise RAII (see the CommandMacro class for an example).

There are two methods to implement this functionality:

1. Move the cleanup functionality into a function and call it both from within the catch statement and after the catch statement. This is the non-C++ way. Some other language have the finally keyword for this purpose, which is missing from C++, because that is non-idiomtic code in this language.

2. You write a little helper class (can be placed fully into the cpp file) that does the following:
- Do the initial change in the constructor
- Do the cleanup (reverting the change) in the destructor
- Delete the copy constructor and the assignment operator

This is the RAII method and it is the preferred approach in C++.

Reply:
I've been thinking of what variables should have their own class with custom destructors. I haven't found any better ways to do things yet.

Here are all the variables that are defined inside the default functionality of try-catch (except the CSVRender::InstanceMode::dropInstance, which is quite similar):

```in CSVRender::InstanceMode::handleDropMethod
float smallestDropHeight -> deleted when case-scope ends
int counter -> deleted when case-scope ends
osg::ref_ptr<TagBase> -> smart pointer (RAII?)
CSVRender::ObjectTag* -> pointer, deleted when scope ends (and dynamic_cast)
float thisDrop -> deleted when if-scope ends
osg::ref_ptr<TagBase> -> smart pointer (RAII?)
CSVRender::ObjectTag* -> pointer, deleted when scope ends (and dynamic_cast)
ESM::Position -> struct with floats, should be deleted when scope ends ( https://github.com/OpenMW/openmw/blob/dd67cdfa554c2bbd5aa073fa36a6bb736de4cca1/components/esm/defs.hpp#L38 )
objectTag->mObject->apply uses CommandMacro (RAII)

in CSVRender::InstanceMode::getDropHeight
const osg::Vec3d& point -> deleted when function-scope ends?
osg::Vec3d -> deleted when function-scope ends?
osg::ref_ptr<osgUtil::LineSegmentIntersector> -> smart pointer (RAII?)
osgUtil::IntersectionVisitor -> deleted when function-scope ends?
osgUtil::LineSegmentIntersector::Intersections::iterator it -> deleted when for-scope ends```

However, if something results in an exception, and the clean-up you are referring to is about the change of a mask, perhaps something like the code below would be the correct solution?

```
void CSVRender::InstanceMode::myCustomHelperClassChangesMask::myCustomHelperClassChangesMask()
{
    CHANGE MASK HERE
}

CSVRender::InstanceMode::myCustomHelperClassChangesMask::~myCustomHelperClassChangesMask()
{
    CHANGE MASK BACK HERE
}

void CSVRender::InstanceMode::handleDropMethod(DropMode dropMode, QString commandMsg)
{
    myCustomHelperClassChangesMask()
    try
    {
         DROPPING FUNCTIONALITY HERE
    }
    catch (const std::exception& e)
    {
        throw;
    }
} // When this scope ends, the destructor of myCustomHelperClassChangesMask is called```

Reply:
edit note: Removed the `void` etc. crap from previous meta-code example...

Reply:
> I've been thinking of what variables should have their own class with custom destructors.

It's possible to customize type by a callback using templates, `std::function` and lambdas:
```c++
template <class OnDestruct>
struct MyType {
    OnDestruct onDestruct;
    ~MyType() { onDestruct(this); }
};

template <class OnDestruct>
MyType<OnDestruct> makeMyType(OnDestruct&& onDestruct) {
    return MyType<std::decay_t<OnDestruct>> {std::forward<OnDestruct>(onDestruct)};
}

auto foo = makeMyType([] (auto self) { /*...*/ });
auto bar = makeMyType([] (auto self) { /*...*/ });
```

Also there is such generic type called `std::unique_ptr`.
```
struct MyType {};

MyType foo;
auto onFooDestruct = [] (auto self) {  };
std::unique_ptr<MyType, decltype(onFooDestruct)> fooWithCustomDestructor(&foo, onFooDestruct);
```

Reply:
@unelsson The code you posted looks nearly correct. You can delete the try-catch-clause here, because if the only thing your catch does is to re-throw, then it becomes a no-op.

@elsid Interesting approach, but probably a bit more than we need for this rather simple problem.

Reply:
Thanks for your great reviews. I used the constructor-destructor data handling method (made a custom helper class) and rebased this PR to the latest master.

Reply:
Functionally this looks good. There are a few stylistic issues. I will comment about them inline. Once these issues have been addressed this PR should be ready for a merge.